### PR TITLE
Yards search

### LINKED ISSUE #27: Get yards that match search criteria
/api/v1/yards?location=ZIP&purposes=PURPOSE+NAME&OTHER+PURPOSE+NAME

- [x] the location query parameter is required
- [x] purposes are optional

### PR DESCRIPTION
**Pull request by Gvieve**
### Add
* yards search route and controller action for query parameters
* search by zipcode and purpose methods on yards model
* purpose name is downcased on the model when saved
### Testing
* happy/sad path and edge case testing for yards search
* model tests for new methods on yard and purpose

Closes #27

### CODE REVIEW

File: APIContract.md [hunk 1]
```diff
@@ -252,7 +252,8 @@
   ```
 
 ## Yard Search
-`GET /api/v1/yards?location=19125&purposes=pet+rental&pet+rental`
+ * The location query parameter is required.
+`GET /api/v1/yards?location=19125&purposes[]=pet+rental&purposes[]party+rental`
   ```json
```
Comment (amsmyth1) [inline APIContract.md:257]: THANK YOU for being so diligent with this. The frontend will be happy

File: app/controllers/api/v1/yards/search_controller.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  private
+
+  def validate_search_params
+    if params[:location].nil?
+      error = "Please enter a zipcode to search available yards"
+      render_error(error)
+    elsif !params[:location].scan(/\D./).empty? || params[:location].length != 5
+      error = "Invalid zipcode"
```
Comment (amsmyth1) [inline app/controllers/api/v1/yards/search_controller.rb:21]: what does the scan method do, would love to hear for my own education

File: app/models/purpose.rb [hunk 3]
```diff
@@ -3,4 +3,10 @@ class Purpose < ApplicationRecord
   has_many :yards, through: :yard_purposes
 
   validates_presence_of :name
+
+  before_save :downcase_name
+
```
Comment (amsmyth1) [inline app/models/purpose.rb:8]: love this. i never think to use before save methods, I definitely need to start incorporating.

File: spec/models/purpose_spec.rb [hunk 4]
```diff
@@ -8,5 +8,13 @@
 
   describe 'validations' do
     it { should validate_presence_of :name }
+
+    describe 'downcase name' do
+      it "downcases the name before save" do
+        purpose = Purpose.create!(name: 'SIGH')
+
+        expect(purpose.name).to eq('sigh')
+      end
```
Comment (amsmyth1) [inline spec/models/purpose_spec.rb:17]: siiiiigggggghhhh

**Review by amsmyth1 — APPROVED** [review]
Great work! This is a great foundation so the front end can be fleshed out. What do you think about creating dummy data? Right now we only ever get the sad paths when we test the endpoints ...

File: app/models/purpose.rb (hunk 3 above)
Comment (Gvieve) [reply app/models/purpose.rb:8]: Thank you!

File: app/controllers/api/v1/yards/search_controller.rb (hunk 2 above)
Comment (domo2192) [reply app/controllers/api/v1/yards/search_controller.rb:21]: I love this too! If you want to be lazy https://github.com/dgilperez/validates_zipcode lol super clean gem

Comment (Gvieve) [reply app/controllers/api/v1/yards/search_controller.rb:21]: It scans a string for the argument given. In this case it's a regex (regular expression) that's looking for any non-numerical characters, if found it puts those results into an array. So, if it's not empty it means there are only numerical characters in the value.

File: spec/requests/api/v1/yard_search_spec.rb [hunk 5]
```diff
@@ -2,86 +2,160 @@
...
+    end
+
+    it "returns an error when the zipcode is not 5 numeric characters" do
+      get "/api/v1/yards/yard_search?location=1346"
+
+      expect(response).to_not be_successful
+      json = JSON.parse(response.body, symbolize_names:true)
+
+      expect(response).to have_http_status(:bad_request)
+      expect(json[:error]).to be_a(String)
```
Comment (domo2192) [inline spec/requests/api/v1/yard_search_spec.rb:157]: Great job being so thorough with testing! PM Ian will appreciate

**Review by domo2192 — APPROVED** [review]
